roboscreenshots/src/main/java/com/google/android/horologist/screenshots/ScreenshotTestRule.kt

```
@@ -0,0 +1,236 @@
+
+import android.app.Application
+import android.content.res.Resources
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.view.View
+import androidx.compose.foundation.background
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.shape.CircleShape
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.CompositionLocalProvider
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.platform.LocalView
+import androidx.compose.ui.test.SemanticsNodeInteraction
+import androidx.compose.ui.test.junit4.ComposeContentTestRule
+import androidx.compose.ui.test.junit4.createComposeRule
+import androidx.compose.ui.test.onRoot
+import androidx.test.core.app.ApplicationProvider
+import androidx.wear.compose.material.Scaffold
+import androidx.wear.compose.material.TimeText
+import coil.compose.LocalImageLoader
+import com.google.android.horologist.annotations.ExperimentalHorologistApi
+import com.google.android.horologist.compose.tools.coil.FakeImageLoader
+import com.google.android.horologist.screenshots.a11y.A11ySnapshotTransformer
+import com.quickbird.snapshot.Diffing
+import com.quickbird.snapshot.FileSnapshotting
+import com.quickbird.snapshot.Snapshotting
+import com.quickbird.snapshot.fileSnapshotting
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.runTest
+import org.junit.rules.RuleChain
+import org.junit.rules.TestName
+import org.junit.rules.TestRule
+import org.junit.runner.Description
+import org.junit.runners.model.Statement
+import org.robolectric.RobolectricTestRunner
+
+/**
+ * A [TestRule] that allows you to run screenshot tests of your composable components.
+ *
+ * This rule requires robolectric, so the test suite should run with [RobolectricTestRunner]. See
```

I wonder if we can test this via the Rule api?

---

you mean add the RunWith, Config and GraphicsMode via Rule? that would be the ideal! will explore it in further iteration

---

I meant something different.  I guess first question, is whether it really does require it? https://robolectric.org/androidx_test/

> It is now possible to use the AndroidX test runner in Robolectric tests. If you require a custom test runner currently

Is this because we request customisation?

But I do think I've had failures because I wasn't using the robolectric runner, and they were not obvious.  I wonder whether the rule can fail fast by just checking which runner is active via the junit APIs?

---

On the second one, you can certainly change the qualifiers during the test.  Not sure about GraphicsMode, it seems a very fundamental early split, based on https://github.com/robolectric/robolectric/issues/8073